diff -uraN uboot-2010.06/include/configs/hi3516cv300.h uboot-2010.06-b/include/configs/hi3516cv300.h
--- uboot-2010.06/include/configs/hi3516cv300.h	2021-08-01 19:42:40.014720097 +0300
+++ uboot-2010.06-b/include/configs/hi3516cv300.h	2021-08-01 19:43:25.422229798 +0300
@@ -319,7 +319,7 @@
 /*-----------------------------------------------------------------------
  * Snapshot boot support
  * ----------------------------------------------------------------------*/
-#define CONFIG_SNAPSHOT_BOOT		1
+//#define CONFIG_SNAPSHOT_BOOT		1
 #ifdef CONFIG_SNAPSHOT_BOOT
 /* #define CONFIG_ARCH_MMU */ /* enable MMU for fast decompress */
 #define CONFIG_SHA1
@@ -350,6 +350,6 @@
 
 /* #define __LITTLE_ENDIAN	1 */
 
-#define CONFIG_CMD_UGZIP 1
+//#define CONFIG_CMD_UGZIP 1
 
 #endif	/* __CONFIG_H */
diff -uraN uboot-2010.06/include/configs/hi3516ev100.h uboot-2010.06-b/include/configs/hi3516ev100.h
--- uboot-2010.06/include/configs/hi3516ev100.h	2021-08-01 19:42:40.014720097 +0300
+++ uboot-2010.06-b/include/configs/hi3516ev100.h	2021-08-01 19:43:38.146092242 +0300
@@ -321,7 +321,7 @@
 /*-----------------------------------------------------------------------
  * Snapshot boot support
  * ----------------------------------------------------------------------*/
-#define CONFIG_SNAPSHOT_BOOT		1
+//#define CONFIG_SNAPSHOT_BOOT		1
 #ifdef CONFIG_SNAPSHOT_BOOT
 /* #define CONFIG_ARCH_MMU */ /* enable MMU for fast decompress */
 #define CONFIG_SHA1
@@ -352,6 +352,6 @@
 
 /* #define __LITTLE_ENDIAN	1 */
 
-#define CONFIG_CMD_UGZIP 1
+//#define CONFIG_CMD_UGZIP 1
 
 #endif	/* __CONFIG_H */
